[1rbg] iP

1rbg opened the pull request:
### CoDriver
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

CoDriver is a new chatbot, running GPT(-1). It's 
- text-based
- easy to learn
- ~~FAST~~ _Super_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/1rbg/ip/releases/tag/A-Jar).
2. launch it in terminal using `java -jar codriver.jar`.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

------------------------------------------------------------
If you are a Java programmer, you can use it to practice Java too. Here's the main method:
``` Java
public class Main {
	public static void main(String[] args) {  
	    new CoDriver("./data/codriver.txt").run();  
	}
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

1rbg committed:
Level-0

1rbg committed:
Implemented echo functionality

1rbg committed:
Revert "Implemented echo functionality"

This reverts commit 7c39a479ac71385d284eff9e1305c54f90c61525.

1rbg committed:
Implemented echo functionality

1rbg committed:
Added ability to store text and display

1rbg committed:
Implemented mark as done and undone functionality

1rbg committed:
Shifted printSepLine() to it's own class

1rbg committed:
Changed Task to abstract class

1rbg committed:
Subclasses of Task

1rbg committed:
Improved CoDriver interaction with user

1rbg committed:
Implemented ToDo, Event, Deadline

1rbg committed:
Added simple single test

1rbg committed:
Added new exception for erroneous commands

1rbg committed:
Implemented error handling for erroneous commands

1rbg committed:
Changed test script

1rbg committed:
Implemented ArrayList instead of array, added exceptions for mark and unmark

1rbg committed:
Implemented ArrayList instead of array, added exceptions for mark and unmark

1rbg committed:
Changed TaskList constructor, added exceptions for mark and unmark

1rbg committed:
Added support for deleting tasks from TaskList

1rbg committed:
Added test cases for unmark and mark, and delete

1rbg committed:
Added more error handling cases for mark, unmark, delete

1rbg committed:
Changed test script

1rbg committed:
Consolidate parsing of command into Parser class

To simplify the addition of new commands, a new enum Command is
used. Parser and Command make the code more readable and is easier
to expand.

1rbg committed:
Add save functionality

1rbg committed:
Merge branch 'branch-Level-7'

1rbg committed:
Change date storage

1rbg committed:
Merge branch 'branch-Level-8'

1rbg committed:
Refactor commands into Command abstraction

Creation of new class Command to utilise polymorphism.

Shift all parsing methods from respective classes into Parse,
to keep the parsing in a black box abstraction. CoDriver logic
is updated accordingly.

1rbg committed:
Merge branch 'branch-A-MoreOOP'

1rbg committed:
Remove unused import statements

1rbg committed:
Merge branch 'add-gradle-support' of https://github.com/1rbg/ip into A-Gradle

1rbg committed:
Change mainClassName to CoDriver

1rbg committed:
Merge branch 'branch-A-Gradle'

1rbg committed:
Add preliminary JUnit tests

1rbg committed:
Remove parseDate method from ParserTest

1rbg committed:
Add more test methods

1rbg committed:
Merge branch 'branch-A-JUnit'

1rbg committed:
Fix botched name refactor

1rbg committed:
Fix botched name refactor

1rbg committed:
Fix index out of bounds bug

1rbg committed:
Configure shadowJar

1rbg committed:
Merge branch 'branch-A-Jar'

1rbg committed:
Add JavaDoc comments

1rbg committed:
Fix style infractions

1rbg committed:
Add find command

1rbg committed:
Change import statements to alphabetical order

1rbg committed:
Merge branch 'branch-A-CodingStandard'

1rbg committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/lrbg/codriver/command/UnknownCommand.java

1rbg committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/lrbg/codriver/data/TaskList.java
#	src/main/java/lrbg/codriver/parser/Parser.java
#	src/main/java/lrbg/codriver/ui/Ui.java
#	src/test/java/lrbg/codriver/parser/ParserTest.java

PallonCX commented on src/main/java/lrbg/codriver/data/Task.java:
```diff
@@ -0,0 +1,64 @@
+     * Returns a string representation of the task to be saved in a file.
+     * @return The string representation of the task to be saved.
+     */
+    public String toSaveString() {
```
The lecture topic for Week 4 content suggests that:

> A name is not just for differentiation; it should explain the named entity to the reader accurately and at a sufficient level of detail.

Considering the method in line 52 is called `toString`, which generates a `String` as its return value, a reader may naturally interpret it as **a function that describes the object as text**.

However, the method name for line 60 is called `toSaveString`. A reader may not have a clear idea about "what a SaveString is", or may interpret the function as **a function that saves a string somewhere**. These two interpretations are clearly incorrect if the reader reviews the method implementation.

Therefore as a suggestion, perhaps it will be good to give the function a better name? One of the example that comes to my mind is `toStringForSaveInFile`.

PallonCX commented on src/main/java/lrbg/codriver/data/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+     */
+    public TaskList(String[] taskStrings) {
+        this.tasks = new ArrayList<>();
+        for (String s : taskStrings) {
+            String[] arguments = s.split("\\|");
+            String type = arguments[0];
+            boolean isDone = arguments[1].equals("1");
+            switch (type) {
+            case "T": // task.ToDo
+                String description = arguments[2];
+                Todo t = new Todo(description);
+                if (isDone) {
+                    t.markDone();
+                }
+                this.tasks.add(t);
+                break;
+            case "D": // task.Deadline
+                String deadlineDescription = arguments[2];
+                LocalDate deadlineDate = Parser.parseDate(arguments[3]);
+                Deadline d = new Deadline(deadlineDescription, deadlineDate);
+                if (isDone) {
+                    d.markDone();
+                }
+                this.tasks.add(d);
+                break;
+            case "E": // task.Event
+                String eventDescription = arguments[2];
+                String[] dates = arguments[3].split("~");
+                LocalDate eventFrom = Parser.parseDate(dates[0]);
+                LocalDate eventTo = Parser.parseDate(dates[1]);
+                Event e = new Event(eventDescription, eventFrom, eventTo);
+                if (isDone) {
+                    e.markDone();
+                }
+                this.tasks.add(e);
+                break;
+            }
+        }
```
The lecture topic for Week 4 content suggests that:

> A name is not just for differentiation; it should explain the named entity to the reader accurately and at a sufficient level of detail.

Noticed that within these part of codes, you have used `s`, `t`, `d`, and `e` as your variable name. Perhaps name formed by single character is more suitable for temporary usage such as loop variable, but for here I think giving a more complete variable name can effectively improve the readability of code.

meenulekha commented on src/main/java/lrbg/codriver/command/ByeCommand.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Creates a new ByeCommand.
+     */
+    public ByeCommand() {
```
I think the default constructor will be used even if you don't explicitly include one!

PallonCX commented on src/main/java/lrbg/codriver/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+
+            StringBuilder descriptionBuilder = new StringBuilder();
+            int i;
+            for (i = 1; i < arguments.length; i++) {
```
The Java Coding Standard that we are referring suggests that:

> Variables should be initialized where they are declared and they should be declared in the smallest scope possible.

Hence, perhaps you may want to modify the code as such:
```Suggestion
            int i = 1;
            for (; i < arguments.length; i++) {
```

meenulekha commented on src/main/java/lrbg/codriver/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+                throw new CoDriverException("Error! You must provide a /from date for an event!");
+            }
+
+//        StringBuilder fromBuilder = new StringBuilder();
```
make sure your comments are aligned with the rest of the code:)

meenulekha commented on src/main/java/lrbg/codriver/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+                if (fromDate == null) {
+                    throw new CoDriverException("Error! The date provided must be in yyyy-mm-dd format!");
+                }
+//            fromBuilder.append(arguments[i]).append(" ");
```
you can remove any comments of code that is no longer being used :)

PallonCX commented on src/main/java/lrbg/codriver/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+                throw new CoDriverException("Error! You cannot provide an event with no description!");
+            }
+            descriptionBuilder.deleteCharAt(descriptionBuilder.length() - 1); // remove the last space
+
```
The Java Coding Standard that we are referring suggests that:

> Logical units within a block should be separated by one blank line.

Hence, perhaps you may want to insert some blank lines to improve the code readability?

PallonCX reviewed (commented):
Well done! LGTM. Just some nits to fix.

tanguoyang commented on src/main/java/lrbg/codriver/command/Command.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return True if the command is an exit command.
+     */
+    public boolean isExit() {
```
As the list of commands grow, there will only ever be 1 exit command. I wonder if theres a better way to structure the code instead of having every command have isExit() = false?

tanguoyang commented on src/main/java/lrbg/codriver/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * {@inheritDoc}
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws CoDriverException {
```
I wonder if theres a better name for CoDriverException? CoDriverException doesnt tell me much about what kind of error is happening

tanguoyang commented on src/main/java/lrbg/codriver/command/DeleteCommand.java:
```diff
@@ -0,0 +1,48 @@
+     * @param obj The object to compare to.
+     * @return True if the given object contains the same index, false otherwise.
+     */
+    public boolean testEquals(Object obj) {
```
consider overriding equals() and hashcode() function instead!

tanguoyang commented on src/main/java/lrbg/codriver/command/EventCommand.java:
```diff
@@ -0,0 +1,40 @@
+        ui.showAddTask(newTask, tasks.size());
+    }
+
+    public boolean testEquals(Object obj) {
```
i wonder if testEquals is better placed somewhere else, like a CommandTest class built for testing instead?

tanguoyang commented on src/test/java/lrbg/codriver/parser/ParserTest.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Test
+    public void parseTest_deadlineIncorrectArgument_exceptionThrown() {
```
Might be better if our deadlineIncorrectArgument has a more specific case. There are many ways in which the argument is incorrect. when i run the test in my terminal, i should be able to know what kind of incorrect argument throws an error!

tanguoyang commented on src/test/java/lrbg/codriver/parser/ParserTest.java:
```diff
@@ -0,0 +1,158 @@
+ */
+public class ParserTest {
+    @Test
+    public void parseTest_correctArgument_success() {
```
Might be better if our correctArgument has a more specific case. There are many ways in which the argument is correct. when i run the test in my terminal, if it fails i should be able to look at the test name and quickly come up with a diagnosis!

tanguoyang reviewed (commented):
minor comments, keep up the good work!

triciiaaa commented on src/main/java/lrbg/codriver/data/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+                }
+                this.tasks.add(e);
+                break;
+            }
```
Would it be better to include a default branch in `case` statements to ensure that all possible outcomes have been considered at the branching point?

triciiaaa commented on src/main/java/lrbg/codriver/data/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+     * Gets the size of the list.
+     * @return The size of the list.
+     */
+    public int size() {
+        return this.tasks.size();
+    }
```
Would it be better to rename the method as `getSize()` instead?

triciiaaa commented on src/main/java/lrbg/codriver/data/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+     * @param index The index of the task, zero indexed.
+     * @return The task at the given index.
+     */
+    public Task get(int index) {
+        return this.tasks.get(index);
+    }
```
In addition to the comment on the `size()` method, would it be better to rename this method as `getTask(int index)`?

triciiaaa commented on src/main/java/lrbg/codriver/data/Event.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class Event extends Task {
+    /** The start date of the event. */
+    private final LocalDate from;
+    /** The end date of the event. */
+    private final LocalDate to;
```
Perhaps you may consider renaming the variables `from` and `to` to `startDate` and `endDate` respectively to make it more intuitive for the user?

triciiaaa reviewed (commented):
Overall, I found your code easy to read and I like how you separated the different commands into different classes! Keep it up 👍

1rbg committed:
Add GUI functionality

CoDriver only had a CLI.

JavaFX allows us to implement a GUI that can sit on top of the CLI
that exists.

Ui class refactored to return String for all show methods, and
execute method Command class also refactored to return String for
compatibility with the JavaFX code.

1rbg committed:
Merge branch 'branch-Level-10'

1rbg committed:
Add assertions

1rbg committed:
Refactor for code quality

Some variable names are misleading / unclear.

Changing variable names to improve clarity will improve code
readability.

1rbg committed:
Set jar to launch GUI

1rbg committed:
Refactor code for readability

parse() method contains too many layers of abstraction and is too
long.

Extracting the methods to reduce one layer of abstraction makes the
code more readable and easier to navigate.

1rbg committed:
Refactor code

Made the happy path more prominent

1rbg committed:
Merge pull request #1 from 1rbg/branch-A-CodeQuality

A-CodeQuality

1rbg committed:
Merge branch 'master' into branch-A-Assertions

1rbg committed:
Merge pull request #2 from 1rbg/branch-A-Assertions

A-Assertions

1rbg committed:
Add ability to detect duplicate tasks

1rbg committed:
Merge branch 'branch-C-DetectDuplicates'

1rbg committed:
Add titlebar to GUI

1rbg committed:
Fix bug from refactoring

1rbg committed:
Add product screenshot image

1rbg committed:
Add UG and features to README

1rbg committed:
Add file saving GUI and README

1rbg committed:
Rename UI.png to Ui.png
